File: solution/02xxx/2504 - 괄호의 값/2504.c
#include <stdio.h>
#include <string.h>

int sum, idx = -1, temp = 1;
char stack[33], str[33];

char top() {return stack[idx];}
void push(char a) {stack[++idx] = a; return;}
void pop() {stack[idx--] = 0; return;}
int empty() {if(idx < 0) return 1; else return 0;}

int main() {
    scanf("%s", str);
    int len = strlen(str);
    for(int i = 0; i < len; i++) {
        if(str[i] == '(') push('(');
        else if(str[i] == '[') push('[');
        else if(str[i] == ')' && !empty() && top() == '(') pop();
        else if(str[i] == ']' && !empty() && top() == '[') pop();
        else {
            printf("%d", sum);
            return 0;
        }
    }
    if(!empty()) {
        printf("0");
        return 0;
    }
    for(int i = 0; i < len; i++) {
        if(str[i] == '(') {
            temp *= 2;
            push('(');
            if(i+1 < len && str[i+1] == ')') {
                sum += temp;
                temp /= 2;
                pop();
                i++;
            }
        }
        else if(str[i] == '[') {
            temp *= 3;
            push('[');
            if(i+1 < len && str[i+1] == ']') {
                sum += temp;
                temp /= 3;
                pop();
                i++;
            }
        }
        else if(str[i] == ')') {
            pop();
            temp /= 2;
        }
        else if(str[i] == ']') {
            pop();
            temp /= 3;
        }
    }
    printf("%d", sum);
}

File: solution/02xxx/2504 - 괄호의 값/2504.py
sum = 0
idx = -1
temp = 1
stack = []

str = input()
strlen = len(str)

for i in range(strlen):
    if str[i] == '(' or str[i] == '[':
        stack.append(str[i])
    elif str[i] == ')' and len(stack) != 0 and stack[-1] == '(':
        stack.pop()
    elif str[i] == ']' and len(stack) != 0 and stack[-1] == '[':
        stack.pop()
    else:
        print(sum)
        exit()

if len(stack) != 0:
    print("0")
    exit()

i = 0
while i < strlen:
    if str[i] == '(':
        temp = temp*2
        stack.append('(')
        if i+1 < strlen and str[i+1] == ')':
            sum = sum+temp
            temp = temp/2
            stack.pop()
            i = i+1
    elif str[i] == '[':
        temp = temp*3
        stack.append('[')
        if i+1 < strlen and str[i+1] == ']':
            sum = sum+temp
            temp = temp/3
            stack.pop()
            i = i+1
    elif str[i] == ')':
        stack.pop()
        temp = temp/2
    elif str[i] == ']':
        stack.pop()
        temp = temp/3
    i = i+1

print(int(sum))